### ISSUE TITLE
Feature: Native Cloud Workspaces (Web IDE) 

### ISSUE THREAD

**Issue by ExplodingDragon** [go-gitea/gitea#27766]
*Note: This is a draft proposal for discussion. Previous iterations are available in the edit history.*

# Feature: Native Cloud Workspaces

## 1. Motivation
Enable users to launch development environments directly from Gitea using a browser or local IDE. 

This design proposes a decoupled architecture to keep the Gitea core lightweight. It delegates infrastructure orchestration to Regional Agents and relies on standard cluster networking for instance management, supporting both declarative (`devcontainer.json`) and zero-config (default base image) environments.

---

## 2. Architecture Overview

```mermaid
graph TD
    subgraph Client ["Client Layer"]
        Browser["Browser (Web IDE / Port Preview)"]
        SSH["Local Terminal (SSH)"]
    end

    subgraph ControlPlane ["Gitea Control Plane"]
        Gitea["Gitea (Metadata & Auth)"]
    end

    subgraph AgentLayer ["Regional Agent (Gateway)"]
        Gateway["Traffic Gateway (Proxy & Router)"]
        Scheduler["Scheduler (K8s/Incus API)"]
    end

    subgraph InstanceLayer ["Workspace Instance"]
        TinyAgent["Tiny Agent (Optional)"]
        IDE["IDE Server (Default 3000)"]
        App["User App (Dynamic Ports)"]
        Storage[("Workspace Volume")]
    end

    %% External Traffic
    Browser -->|"HTTPS / WSS"| Gateway
    SSH -->|"SSH"| Gateway
    Gateway -.->|"Token Verification & Routing"| Gitea

    %% Control & Scheduling
    Gitea == "gRPC Stream (Control Plane)" ==> Scheduler
    Scheduler -->|"Provision/Destroy/Query API"| InstanceLayer

    %% Data Routing
    Gateway == "Direct IP Route / Optional Tunnel" ==> InstanceLayer
    Storage --- InstanceLayer
```

---

### 3.1 Gitea
* **State Machine:** Manages workspace lifecycle (Pending, Running, Stopped, Terminated) and repository bindings.
* **Auth & Tokens:**
    * Issues long-lived gRPC credentials for Regional Agents.
    * Issues short-lived session tokens for client access.
    * Injects a repository-scoped `GITEA_TOKEN` into the workspace for Git operations.
* **Signaling:** Exposes gRPC endpoints to dispatch start/stop commands to Regional Agents.

### 3.2 Regional Agent
A stateless Go daemon deployed near the compute clusters (Kubernetes or Incus).
* **Scheduler:** Translates Gitea commands to infrastructure API calls. Queries cluster APIs directly to retrieve instance status and metrics.
    * *Zero-config Fallback:* If `devcontainer.json` is missing, provisions a predefined base image and injects a `git clone` script to fetch the repository.
* **Gateway:** Listens on a wildcard domain (e.g., `*.ws.gitea.example.com`) and routes external HTTP/SSH traffic directly to the target instance's internal IP.

### 3.3 Optional Tiny Agent
An in-instance daemon is not strictly required. By default, the Regional Gateway routes traffic directly to the container's IP and uses cluster exec APIs (like `kubectl exec`) for terminal sessions.

A Tiny Agent can be optionally deployed inside the workspace for specific edge cases:
* **Strict Network Boundaries:** Establishing an outbound reverse tunnel if the cluster network prohibits direct ingress from the Gateway.
* **Custom SSH Handling:** Bridging SSH traffic without relying on cluster exec APIs.

---

## 4. On-Demand Port Forwarding
Active port discovery inside the container is not required. 

When a user requests a specific port via the assigned wildcard URL (e.g., `https://8080-ws123.ws.gitea.example.com`), the Regional Gateway proxies the traffic directly to port `8080` on the instance's internal IP. If the port is actively listening, the connection succeeds; otherwise, the gateway returns a standard connection error.

---

### 5.1 Compute & Permissions
* **No Privileged Containers:** Force `privileged: false`.
* **Passwordless Sudo (Zero-config):** Base images run as a non-root user but have passwordless `sudo` enabled to allow package installation (`apt install`) without compromising host security.
* **Drop Capabilities:** High-risk Linux capabilities (`CAP_SYS_ADMIN`, `CAP_NET_ADMIN`) are stripped to limit the impact of a `sudo` escalation.

### 5.2 Network
* **Egress Blocking:** Network Policies deny outbound traffic by default, whitelisting only Gateway IPs and public package registries.
* **Anti-Scanning:** Packets destined for internal subnets (`10.0.0.0/8`, `192.168.0.0/16`) are dropped at the network level.

### 5.3 Resource Constraints
* **Token Scoping:** The injected `GITEA_TOKEN` is revoked upon workspace hibernation.
* **Quotas:** Cgroups enforce hard limits on CPU cores and Memory.
* **Idle Timeout:** The Regional Agent monitors traffic and cluster metrics. Idle instances are stopped automatically (e.g., after 30 minutes) to reclaim compute resources.

---

## 6. Storage Model

The storage model separates the ephemeral environment from persistent user data:

| Mount Path | Storage Type | Lifecycle |
| :--- | :--- | :--- |
| `/` | **Root FS** | **Ephemeral.** Built from `devcontainer.json` or the default image. System-level changes persist only while the instance is running. Destroyed upon hibernation. |
| `/workspaces` | **Data Volume** | **Persistent.** Backed by cluster persistent volumes. Contains cloned source code, local build artifacts, and dependencies. Survives hibernation. Deleted only via the Gitea UI. |

**Comment by ExplodingDragon** [comment]
Confirmation is required before starting:

- [ ] Gitea's routing support for WebSocket
- [ ] embedded SSH Server supports port forwarding and sftp. (This way we can use `vscode desktop` or `intellij`.)
- [ ] How to configure permissions for instances that are in a group.

**Comment by silverwind** [comment]
I think this shouldn't depend on act_runner.

**Comment by dydhyhwu** [comment]
> I think this shouldn't depend on act_runner.

I agree with your idea. I think it would be great to consider implementing a built-in Web IDE scheduling mechanism similar to act_runner, which supports features like SSH, VSCode Remote, JetBrains Gateway, etc. It would also be beneficial to enhance the extensibility of the system, allowing integration with mature third-party products such as [Coder](https://coder.com/docs/v2/latest).

**Comment by anbraten** [comment]
FYI There are currently two IDE systems working with Gitea already:
- https://github.com/daytonaio/daytona (support added in https://github.com/daytonaio/daytona/pull/194)
- https://github.com/loft-sh/devpod (works with local git folders)

**Comment by dydhyhwu** [comment]
so cool!
